coding_guidelines.md

```diff
@@ -0,0 +1,500 @@
+
+```typescript
+// DON'T MAKE A CLASS LIKE THIS!
+class Wallet {
+  private _amount: number;
+  private _loan: number = 0;
+  
+  constructor(amount: number) {
+    this._amount = amount;
+  }
+
+  pay(price: number): boolean {
+    if(this._amount - price > 0) {
+      this._amount -= price;
+      return true;
+    } else if(this._loan < 5000 && this._amount !== 0) {
+      const loanAmount = price - this._amount;
+      this._amount = 0;
+      loan += loanAmount;
+      return true;
+    } else if(this._loan < 5000 && this._amount === 0) {
+      loan += price;
+      return true;
+    }
+    return false;
+  }
+}
+```
+> **Note:** In addition to the naming convention this price doesn't check if the price number is > 0 to verify that they aren't adding funcs into the wallet buy "paying" for negative values.
+
+The above example illustrates in a way that the function name is no longer discriptive of what the code does. It also starts to blend into a section covered later [Single Responsibility Pattern](#single-responsibility-pattern), but first let's cover the naming convention. A way to make the function more discriptive would be:
+```typescript
+...
+payFromWalletOrAttemptToLendMoneyToBuy(price: number): boolean {
+  ...
+}
+...
+```
+Now the function name mirrors to an extend the behavior of the function, but in a way the function is still a bit wrong. We return a boolean to see if the purchase was accepted, so a better name might have been.
+```typescript
+...
+successfullyPaidFromWalletOrLentMoneyToBuy(price: number): boolean {
+  ...
+}
+...
+```
+Now in addition to describing what happens inside the function it also matches with the return type of the function.
+
+# Arrays
+
+### [Array Instantiation](#array-instantiation)
+For array instantiation you should use the literal syntax instead of `new` constructor.
+```typescript
+// const items: number[] = new Array(); Don't do this.
+const items: number[] = [];
+```
+
+### [Array Operations](#array-operations)
+For operations on arrays use push etc instead of direct assignment.
+```typescript
+const items: number[] = [];
+// items[items.length] = 14; No, don't do this.
+items.push(14);
+```
+
+# Functions
+
+### [Expression Vs Declaration](#expression-vs-declaration)
+When defining a function we prefer function declaration over function expression, and when dealing with assigned functions we prefer using using an anonymous [arrow function](#arrow-functions) expression.
+```typescript
+function helloWorld(): void {} // Yes
+const helloWorld = function(): Void => {}; // No
+const helloWorld = (): void => {}; // No
+```
+> **Note:** Performance based on numbers from: https://jsperf.com/function-vs-arrow-function/14.
```

Comment:
I don't think that test is conclusive - my results vary with the choice of browser

Reply:
Sure, I just mean whatever codestyle we go for I would prefer we have 1, and not 2.